### PR TITLE
fix(hogvm): unify int and float equality

### PR DESCRIPTION
**Pull request by matheus-vb**
## Problem

A behavioral cohort or an error tracking rule that compares a number misses values stored as the other numeric type.

- A rule reading `properties.code == 1` does not match a recorded `1.0`.
- The Rust HogVM derived `PartialEq` on `Num`, so `Integer(1)` and `Float(1.0)` never compared equal.
- Both reference VMs disagree. TypeScript compares f64 with `===`, and Python uses native `==`.
- The Rust VM also disagreed with itself. Its ordering opcodes already widen a mixed pair to f64, so `1 >= 1.0` was true while `1 == 1.0` was false.
- This blocks the seeder column projection, where ClickHouse `JSONExtractKeysAndValuesRaw` rewrites `1.0` as `1`.
- A backfill would evaluate an integer where the realtime path saw a float, and diverge on every equality leaf.

## Changes

- `Eq`, `NotEq`, `in`, `notIn`, `has`, `indexOf`, `equals` and `notEquals` now treat `1` and `1.0` as the same number.
- A manual `PartialEq for Num` replaces the derive. Two integers compare as i64, and every other pairing widens to f64.
- The new behavior is not gated behind `coerce_comparisons`, so cymbal rule evaluation gets it too. Gating it would leave the default path disagreeing with the references.
- A mixed pair above 2^53 rounds when it widens. That matches TypeScript, diverges from Python, and a test pins it as deliberate.
- Mechanical: five doc comments corrected, and one stale corpus count dropped.

## How did you test this code?

- `common/hogvm/__tests__/numericEquality.hog` is a new corpus fixture that `tests/parity.rs` diffs against the Node oracle.
- It fails before the change and passes after. No existing corpus program covered an int against float comparison.
- Unit tests pin the three ways this impl can go wrong. NaN stays unequal, two large integers stay exact past 2^53, and a mixed pair past 2^53 rounds as documented.
- VM level tests assert every case on the coercing path and on the default path that cymbal runs, because the change is unconditional.
- Node, Python and compiled JavaScript each produce the committed `.stdout` independently, so the snapshot is verified rather than blessed.
- Not run: the cymbal suites that need the dev stack database.

## Docs update

None.

### CODE REVIEW

File: rust/common/hogvm/src/values.rs [hunk 1]
```diff
@@ -26,12 +26,31 @@ pub struct Upvalue {
...
+// `1 == 1.0` is true. Equality must therefore unify the variants. A mixed pair widens to f64 and
+// compares with IEEE `==`, which is what `Num::binary_op` does for `Lt`/`Gt`, so `a == b` stays
+// consistent with `!(a < b) && !(a > b)`. Two integers compare as i64, so values beyond 2^53 stay
+// exact against each other. A mixed pair above 2^53 rounds like the TS reference; Python compares
+// mixed pairs exactly and can disagree there.
+//
+// Do not assume `Num::compare` agrees. It widens too, but then applies `total_cmp`, a total order
+// that ranks NaN and separates `-0.0` from `0`. IEEE `==` instead keeps NaN unequal to NaN and
+// calls `0` and `-0.0` equal, so the two disagree on those two inputs.
+impl PartialEq for Num {
```
Comment (haacked) [inline rust/common/hogvm/src/values.rs:49]: `question`: Mixed integer and float equality can violate `PartialEq`'s transitivity above 2^53:

```text
a = Integer(9007199254740993)
b = Float(9007199254740992.0)
c = Integer(9007199254740992)

a == b
b == c
a != c
```

Mixed pairs widen to `f64`, so `a` rounds to the same value as `b`. Integer pairs compare as `i64`, so `a` remains different from `c`.

The current semantics still look like the right tradeoff. Always widening would treat two distinct large integer IDs as equal. Comparing mixed pairs exactly would disagree with the ordering operations this PR aligns with. Could we add one line next to the `Num::compare` warning that says equality is not transitive across the integer and float boundary at this magnitude? Without that caveat, future uses of `dedup`, `sort_by`, `binary_search`, or a hash set could produce results that depend on input order.

File: rust/common/hogvm/tests/numeric_coercion.rs [hunk 2]
```diff
@@ -63,6 +68,160 @@ fn array(values: &[Vec<Value>]) -> Vec<Value> {
     bytecode
 }
 
+/// `needle in haystack`. The VM pops the needle first, so the haystack is emitted first.
+fn in_op(needle: &[Value], haystack: &[Value]) -> Vec<Value> {
```
Comment (haacked) [inline rust/common/hogvm/tests/numeric_coercion.rs:72]: `suggestion`: Could `in_op` delegate to [the `compare` helper at line 28](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/tests/numeric_coercion.rs#L28)? `in_op(needle, haystack)` emits exactly the same bytecode as `compare(needle, haystack, OP_IN)`. Both write the `"_H"` header, emit the right operand before the left, then emit the operation and `OP_RETURN`.

Delegating would keep the bytecode header and operand order in one place while preserving the descriptive name and comment:

```rust
/// `needle in haystack`. The VM pops the needle first, so the haystack is emitted first.
fn in_op(needle: &[Value], haystack: &[Value]) -> Vec<Value> {
    compare(needle, haystack, OP_IN)
}
```

The `call` helper below emits its arguments in source order, so its separate implementation should stay.

File: rust/common/hogvm/src/values.rs [hunk 3]
```diff
@@ -929,4 +953,39 @@ mod tests {
             ));
         }
     }
+
+    #[test]
+    fn int_float_equality_unifies() {
+        assert_eq!(Num::Integer(1), Num::Float(1.0));
+        assert_eq!(Num::Float(1.0), Num::Integer(1));
+        assert_ne!(Num::Integer(2), Num::Float(1.0));
+        assert_ne!(Num::Float(1.0), Num::Integer(2));
```
Comment (haacked) [inline rust/common/hogvm/src/values.rs:968]: `suggestion`: Could we add a test that `Integer(0)` equals `Float(-0.0)`? [The caveat above](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/src/values.rs#L42-L44) identifies NaN and signed zero as cases where `Num::eq` and `Num::compare` disagree. [`nan_is_never_equal`](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/src/values.rs#L966) covers NaN, but no test covers signed zero. Hog programs can produce `-0.0` through `0.0 * -1`.

```suggestion
        assert_ne!(Num::Float(1.0), Num::Integer(2));
        assert_eq!(Num::Integer(0), Num::Float(-0.0));
```

File: rust/common/hogvm/src/values.rs [hunk 4]
```diff
@@ -26,12 +26,31 @@ pub struct Upvalue {
...
 pub type UpvalueCell = Rc<RefCell<Upvalue>>;
 
-#[derive(Debug, Clone, PartialEq)]
+#[derive(Debug, Clone)]
 pub enum Num {
     Integer(i64),
     Float(f64),
 }
 
+// The reference VMs treat all numbers as one numeric domain, because JS has only f64 and Python's
```
Comment (haacked) [inline rust/common/hogvm/src/values.rs]: `nit`: Could these `Num` comments use `///` instead of `//`? The comments added in [`vm.rs`](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/src/vm.rs) and at [values.rs:468](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/src/values.rs#L468) currently won't appear in rustdoc. Using `///` would show them under Trait Implementations, as [`memory.rs:30`](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/src/memory.rs#L30) does.

File: rust/common/hogvm/tests/numeric_coercion.rs [hunk 5]
```diff
@@ -63,6 +68,160 @@ fn array(values: &[Vec<Value>]) -> Vec<Value> {
...
+        run_legacy(bytecode).expect("legacy execution succeeds"),
+        expected,
+        "default path: {label}",
+    );
+}
+
+#[test]
+fn int_float_equality_is_numeric_on_both_paths() {
+    // `Integer(1)` and `Float(1.0)` are one value to both reference VMs, and the Rust VM's own
+    // ordering ops already widen mixed pairs, so equality has to agree or trichotomy breaks.
```
Comment (haacked) [inline rust/common/hogvm/tests/numeric_coercion.rs]: `nit`: Could we replace "or trichotomy breaks" with the concrete behavior? `1 == 1.0` would be false even though `1 >= 1.0` is true, as [values.rs:37-38](https://github.com/PostHog/posthog/blob/6af934f9c9cb2d9e03717b1c8548c121689a933d/rust/common/hogvm/src/values.rs#L37-L38) explains.

**Review by haacked — APPROVED** [review]
Nice fix, and the parity work behind it holds up: I traced the surfaces and re-checked the fixture against both reference VMs. One question about transitivity above 2^53 plus a few smaller things inline.

File: rust/common/hogvm/src/values.rs [hunk 6]
```diff
@@ -26,12 +26,31 @@ pub struct Upvalue {
...
+//
+// Do not assume `Num::compare` agrees. It widens too, but then applies `total_cmp`, a total order
+// that ranks NaN and separates `-0.0` from `0`. IEEE `==` instead keeps NaN unequal to NaN and
+// calls `0` and `-0.0` equal, so the two disagree on those two inputs.
+impl PartialEq for Num {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Num::Integer(a), Num::Integer(b)) => a == b,
+            _ => self.to_float() == other.to_float(),
+        }
```
[the remark this answers is not in the document]
Comment (matheus-vb) [reply rust/common/hogvm/src/values.rs:54]: valid

Out of scope though. Membership already coerced before this PR - `Num::from_str("1")` gives Integer(1), so 1 in ['1'] matched while both references rejected it. This diff widens that gap to float-formatted strings, it doesn’t open it.

The suggested fix would flip 1 in ['1'] to false and change results for cohort and cymbal rules that predate this change, so it needs its own PR. I’ve noted the divergence at `HogLiteral::equals` so it isn’t rediscovered.
